src/storage/src/hummock/store/version.rs

```diff
@@ -150,48 +164,33 @@ impl StagingVersion {
         max_epoch_inclusive: HummockEpoch,
         table_id: TableId,
         table_key_range: &'a TableKeyRange,
-    ) -> (
-        impl Iterator<Item = &ImmutableMemtable> + 'a,
-        impl Iterator<Item = &SstableInfo> + 'a,
-    ) {
+    ) -> impl Iterator<Item = &StagingData> + 'a {
         let (ref left, ref right) = table_key_range;
         let left = left.as_ref().map(|key| TableKey(key.0.as_ref()));
         let right = right.as_ref().map(|key| TableKey(key.0.as_ref()));
-        let overlapped_imms = self.imm.iter().filter(move |imm| {
+        // reverse to make the newer data come first.
+        self.data.iter().rev().filter(move |data| {
             // retain imm which is overlapped with (min_epoch_exclusive, max_epoch_inclusive]
-            imm.min_epoch() <= max_epoch_inclusive
-                && imm.table_id == table_id
-                && range_overlap(
-                    &(left, right),
-                    &imm.start_table_key(),
-                    Included(&imm.end_table_key()),
-                )
-        });
-
-        // TODO: Remove duplicate sst based on sst id
-        let overlapped_ssts = self
-            .sst
-            .iter()
-            .filter(move |staging_sst| {
-                let sst_max_epoch = *staging_sst.epochs.last().expect("epochs not empty");
-                sst_max_epoch <= max_epoch_inclusive
-            })
-            .flat_map(move |staging_sst| {
-                // TODO: sstable info should be concat-able after each streaming table owns a read
-                // version. May use concat sstable iter instead in some cases.
-                staging_sst
-                    .sstable_infos
-                    .iter()
-                    .map(|sstable| &sstable.sst_info)
-                    .filter(move |sstable: &&SstableInfo| {
-                        filter_single_sst(sstable, table_id, table_key_range)
-                    })
-            });
-        (overlapped_imms, overlapped_ssts)
+            match data {
+                StagingData::ImmMem(imm) => {
+                    imm.min_epoch() <= max_epoch_inclusive
+                        && imm.table_id == table_id
+                        && range_overlap(
+                            &(left, right),
+                            &imm.start_table_key(),
+                            Included(&imm.end_table_key()),
+                        )
+                }
+                StagingData::Sst(staging_sst) => {
+                    let sst_min_epoch = *staging_sst.epochs.first().expect("epochs not empty");
```

Comment:
Why do we use the max_epoch in staging sst previously? Sounds like a bug. cc @Li0k

Reply:
Sorry for the confusion. I mean using the `max_epoch` previously is a bug. Using `min_epoch` is the correct way.

Reply:
Good Catch, use min_epoch to filter is a better way, even if we don’t have such a usage scenario now